Adds the ability to construct and deconstruct locker locks

Citinited opened the pull request:
This PR adds the ability to construct and deconstruct locker locks.

- Use an airlock electronics on a locker without a lock to add one. The locker will inherit the access from the electronics. You can't lock certain lockers such as cardboard boxes.
- Screwdriver a locker to remove its lock. This will give you back the electronics (emagging destroys the electronics). You can only do this to unlocked lockers.

To create a locker lock:

1. Make sure your locker doesn't have a lock and is closed.
2. Hit the locker with airlock electronics.
3. Wait for construction to complete (20 seconds). The locker's access will be updated with the airlock electronics access.

To deconstruct a locker lock:

1. The locker must be unlocked and closed, and must have either a functional or broken lock.
2. Use a screwdriver on the locker.
3. After 10 seconds (better screwdrivers will be faster), the lock will be removed and the airlock electronics (should they exist) will be dropped at your feet. 

I had to tweak the locker icons slightly - broken lockers now spark instead of a black panel. See below for reference (animation not included)

![image](https://user-images.githubusercontent.com/18459142/41564440-ac76a68e-734a-11e8-81b9-7e1e5e110253.png)

Why? So you can fix emagged / broken lockers, you can construct new lockable lockers, and IMO it makes sense to be able to have this ability.

:cl:
add: You can now use an airlock electronics on a locker to add a lock, and can screwdriver an unlocked locker to remove its lock.
add: You can now remove the locks on broken or emagged lockers.
tweak: Removing the lock from a personal locker now wipes that locker's ID details.
tweak: Broken lockers have had their appearance changed.
/:cl:

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -196,6 +204,68 @@
+	if(!istype(E))
+		return
+	if(doinglockstuff)
+		to_chat(user, "<span class='notice'>Someone is already doing something to [src]!</span>")
```
Sounds weird. I'd go with, like `Someone is currently working on [src]` or something.

Citinited replied:
2lewd4u

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -196,6 +204,68 @@
+		to_chat(user, "<span class='notice'>This locker already has a lock!</span>")
+		return
+	if(broken)
+		to_chat(user, "<span class='notice'>Remove the broken lock first!</span>")
```
do `<b>Unscrew</b> the broken lock first!`

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets/bodybag.dm:
```diff
@@ -46,6 +46,12 @@
 		return 1
 	return 0
 
+/obj/structure/closet/body_bag/handle_lock_addition()
```
All of these functions you wrote literally do the default behavior. What you want to do here is just a `return`, not calling the parent function.

Citinited replied:
The `..()` is here so that you hit the bodybag / cardboard / whatever, instead of it `return`ing and doing nothing. I can change that behaviour if you'd like?

FlattestGuitar replied:
The `..()` is literally redundant if it's here on its own. That is the default behavior of an inherited proc.

Do what you want, just remove the stuff that just states the obvious. :P 

Do hit me up on Discord if you need a deeper explanation, I'm happy to help out.

Citinited replied:
Heh no worries, I think my brain wasn't quite engaged when I did this :P

Quantum-M commented:
I think you could make the lights for a emagged/broken locker be blinking in a erratic fashion and/or releasing sparks.

Can you also explain the process of how one adds/removes a lock from a locker?

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets/secure/genpop.dm:
```diff
@@ -18,6 +18,11 @@
 
 	return ..()
 
+/obj/structure/closet/secure_closet/genpop/handle_lock_addition() //If lock construction is successful we don't care what access the electronics had, so we override it
```
I really don't like the idea that the genpop lockers are inherently different to regular ones. 

I know it's a big ask, but I think it'd be for the best if you enabled their creation using the airlock electronics in some way as well.

Citinited replied:
It was done differently because the genpop lockers handle things with IDs, so it was necessary to check the closet had lockerelectronics before being able to close it. However, I could see if I can add the ability to create genpop lockers somehow...

FlattestGuitar replied:
The idea would be that the genpop lockers have different, special locks. You could still unscrew them and all that jazz, but they'd need a different piece of equipment to create. Not sure what. Perhaps a setting on the airlock electronics? The things aren't any good mass produced.

Citinited replied:
This is the last thing I need to look at before this is merge ready (I think). I'll think of some way to construct genpop lockers

Citinited commented:
@Quantum-M Sounds good, I shall update the sprite accordingly.

Steps to create / remove a locker lock shall be in the OP shortly

Citinited commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -33,11 +33,15 @@
 /obj/structure/closet/Initialize(mapload)
 	if(mapload && !opened)		// if closed, any item at the crate's loc is put in the contents
 		addtimer(CALLBACK(src, .proc/take_contents), 0)
+	if(secure)
```
@FlattestGuitar Would you rather me do initialising the electronics this way, or would you rather me make a bunch of subtypes of `obj/item/electronics/airlock` and setting `lockerelectronics` to those paths?

FlattestGuitar replied:
The current implementation is fine.

Citinited commented:
PR updated with requested changes - I also did a tiny refactor on that `update_icon` proc so it doesn't indent four blocks in

Citinited commented:
More code changes and cleanup. 
- In most areas where I checked `lockerelectronics` I now check `!QDELETED(lockerelectronics)` instead
- Removing a lock from a personal closet now wipes the registered ID
- The personal closet now calls `..()` more often instead of having its own duplicate code for it
- Changed `return 1` and `return 0` to `return TRUE` and `return FALSE`
- Removed `return ..()`

SamHPurp commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -177,25 +211,94 @@
+	if(!istype(S))
+		return
+	if(doinglockstuff)
+		to_chat(user, "<span class='notice'>Someone is already doing something to [src]!</span>")
```
Lets keep this less specific than someone is doing something.

Citinited commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -33,11 +33,15 @@
 	var/anchorable = TRUE
-
+	var/obj/item/electronics/airlock/lockerelectronics //Installed electronics
+	var/doinglockstuff = FALSE //Someone is doing some stuff with the lock here, better not proceed further
```
Done, though I'll keep the comment because it might be of use to someone

Citinited commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -33,11 +33,15 @@
 	var/delivery_icon = "deliverycloset" //which icon to use when packagewrapped. null to be unwrappable.
 	var/anchorable = TRUE
-
+	var/obj/item/electronics/airlock/lockerelectronics //Installed electronics
```
Not a boolean but some stuff does act as if it is a boolean - this var points to the installed lockers so you can check accesses directly from it if you need to etc.

Citinited commented:
I've updated this with feedback. One thing that I noticed was that crates now have their own locker electronics. I think this is fine, given you can't add locks to the crates, but I may allow lock addition to crates.

Citinited commented:
Added crafting recipes for genpop and personal lockers (check OP). This is ready for review

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -384,44 +489,24 @@
-		broken = 1
+		broken = TRUE
 		locked = FALSE
+		if(!QDELETED(lockerelectronics))
```
An emag should break the electronics. Not remove it completely from the world. Same behavior as with emagging an airlock.

Citinited replied:
Emagging an airlock makes you discard the installed electronics when you deconstruct it, though. I can change it to just wipe access from the electronics though

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -177,25 +210,95 @@
+	if(!istype(S))
+		return
+	var/brokenword = broken ? "broken" : null
+	user.visible_message("<span class='notice'>You begin removing the [brokenword] lock on [src]...</span>", "<span class='notice'>[user] begins removing the [brokenword] lock on [src]...</span>")
```
`You begin removing the  lock on...`. This will have a double space between `the` and `lock` for a non-broken lock.

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets/secure/genpop.dm:
```diff
@@ -18,6 +18,11 @@
 	return ..()
 
+/obj/structure/closet/secure_closet/genpop/handle_lock_addition() //If lock construction is successful we don't care what access the electronics had, so we override it
+	if(..())
```
The genpop lockers are still broken as shit. They're not meant to be special in any inherent way; they're just hunks of metal with special electronics attached. The electronics attached should just be a somehow modified version of airlock electronics, not the typepath of the locker. That will be nonsensical.

FlattestGuitar commented on code/game/objects/structures/crates_lockers/closets/secure/personal.dm:
```diff
@@ -4,6 +4,18 @@
 	req_access = list(ACCESS_ALL_PERSONAL_LOCKERS)
 	var/registered_name = null
 
+/obj/structure/closet/secure_closet/personal/examine(mob/user)
```
As with genpop, the personal lockers should be made personal purely because of their electronics, not any special poperties of the metal used to build them. Maybe add a small menu in the airlock electronics that would let you pick what you want to make?

FlattestGuitar commented:
Okay, so we talked a bit about this concept and we've basically agreed that the best way to solve the issue would be to just make personal lockers and genpop lockers impossible to construct and deconstruct. They're a very small portion of all the lockers and it shouldn't be a bit gameplay issue, especially considering neither of them get emagged very often. 

Could you do that? (still terrible sorry for letting this hang for so long)

Citinited commented:
I can do that!

No worries it's taken as long as it has, I've been very busy this week myself so haven't got around to doing much DM coding. I'll have time on the weekend to do this :)

With regards to emagged genpop / personal lockers, do you want the ability to fix the locks available, or would you rather not have that at all?

FlattestGuitar commented:
Might as well leave it as is; just removing the crafting recipes for the weird lockers.

Citinited commented:
Removed genpop / personal locker crafting recipes and fixed your format issue

Citinited commented:
Closing and re-opening due to Travis
